src/Agent.Worker/Release/Artifacts/JenkinsArtifact.cs

```diff
@@ -0,0 +1,163 @@
+            await zipStreamDownloader.DownloadFromStream(
+                downloadedStream,
+                string.IsNullOrEmpty(parentFolder) ? "archive" : string.Empty,
+                parentFolder,
+                localFolderPath);
+        }
+
+        public IArtifactDetails GetArtifactDetails(AgentArtifactDefinition agentArtifactDefinition, IExecutionContext context)
+        {
+            Trace.Entering();
+
+            var artifactDetails =
+                JsonConvert.DeserializeObject<Dictionary<string, string>>(agentArtifactDefinition.Details);
+
+            ServiceEndpoint jenkinsEndpoint = context.Endpoints.FirstOrDefault(e => string.Equals(e.Name, artifactDetails["ConnectionName"], StringComparison.OrdinalIgnoreCase));
+            if (jenkinsEndpoint == null)
+            {
+                throw new InvalidOperationException(StringUtil.Loc("RMJenkinsEndpointNotFound", agentArtifactDefinition.Name));
+            }
+
+            string relativePath;
+            var jobName = string.Empty;
+
+            var allFieldsPresents = artifactDetails.TryGetValue("RelativePath", out relativePath)
+                                    && artifactDetails.TryGetValue("JobName", out jobName);
+            if (allFieldsPresents)
+            {
+                return new JenkinsArtifactDetails
+                {
+                    RelativePath = relativePath,
+                    AccountName = jenkinsEndpoint.Authorization.Parameters[EndpointAuthorizationParameters.Username],
+                    AccountPassword = jenkinsEndpoint.Authorization.Parameters[EndpointAuthorizationParameters.Password],
+                    BuildId = Convert.ToInt32(agentArtifactDefinition.Version, CultureInfo.InvariantCulture),
+                    JobName = jobName,
+                    Url = jenkinsEndpoint.Url
+                };
+            }
+            else
+            {
+                throw new InvalidOperationException(StringUtil.Loc("RMArtifactDetailsIncomplete"));
+            }
+        }
+
+        private static void SetupHttpClient(HttpClient httpClient, string userName, string password)
+        {
+            httpClient.Timeout = Timeout.InfiniteTimeSpan;
+            httpClient.DefaultRequestHeaders.Authorization = CreateBasicAuthenticationHeader(userName, password);
+        }
+
+        private static AuthenticationHeaderValue CreateBasicAuthenticationHeader(string username, string password)
+        {
+            var authenticationHeader = string.Format(
+                CultureInfo.InvariantCulture,
+                "{0}:{1}",
+                username ?? string.Empty,
+                password ?? string.Empty);
+
+            return new AuthenticationHeaderValue(
+                "Basic",
+                Convert.ToBase64String(Encoding.UTF8.GetBytes(authenticationHeader)));
+        }
+
+        private static string GetParentFolderName(string relativePath)
```

Comment:
Consider `System.IO.Path.GetDirectoryName`. Although beware, if the path ends in a slash then it doesn't remove the last segment (just removes the last slash).